Sample: customizing Web Chat with plain UI built from ground up

compulim opened the pull request:
While working on hackathon this week, I built this as a base for projects that use Web Chat without our UI.

By cloning this sample, developers could build their own chatting UI that talk `DirectLineJS`, without including any of our UI components. Developers could build an UI with their own styles and signatures.

This project will also serve as a base project for other "breakaway component" samples, which are more-or-less cherry-picking existing Web Chat UIs and adding to this project.

(@EricDahlvang you might be interested in this.)

corinagum commented on samples/19.customization-plain-ui/package.json:
```diff
@@ -0,0 +1,28 @@
+  "homepage": "https://microsoft.github.io/BotFramework-WebChat/19.customization-plain-ui/",
+  "dependencies": {
+    "botframework-webchat": "*",
+    "botframework-webchat-component": "^*",
```
What's the difference between `^*` and `*`?

compulim replied:
Bug in lerna I guess. We should use `*`.

compulim replied:
Fixed.

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,40 @@
+   - Support `imBack`, `messageBack`, and `postBack`
+   - Support rendering hyperlinks
+
+Although UI is main part of Web Chat, the state machine and business logic in Web Chat core is very valuable for building your highly-customized chat UI.
```
```suggestion
Although the UI is one of the main features of Web Chat, the state machine and business logic in Web Chat core is very valuable for building your highly-customized chat UI. 
```

I would move this to the top of the README.md (ln 3)

corinagum replied:
I feel like more discussion on the many benefits of Web Chat core would be good.

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,40 @@
+
+This project is based on [`create-react-app`](https://github.com/facebook/create-react-app).
+
+The completed code contains multiple files. You can start by reading [`App.js`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/19.customization-plain-ui/src/App.js).
```
```suggestion
The completed code contains multiple files. You can start by reading [`App.js`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/19.customization-plain-ui/src/App.js), which replaces the default Web Chat renderer with PlainWebChat and uses its own UI.
```

corinagum commented on samples/19.customization-plain-ui/src/PostBackButton.js:
```diff
@@ -0,0 +1,15 @@
+
+const PostBackButton = ({ cardAction, sendPostBack }) =>
+  <button
+    // Web Chat do the heavylifting for us by exposing a "sendPostBack" function.
```
```suggestion
    // Web Chat does the heavylifting for us by exposing a "sendPostBack" function.
```

compulim commented:
@corinagum I have fixed all grammars and added a section of layering. Could you take a look at it again? I hope to use this README (or might move it somewhere later) to educate people how to customize Web Chat and why it is more valuable than writing your own component (because you can reuse Web Chat business logic, etc).

corinagum commented on README.md:
```diff
@@ -228,45 +228,45 @@ npm run prepublishOnly
+| [`16.customization-selectable-activity`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/16.customization-selectable-activity)                                                                               | Advanced tutorial: Demonstrates how to add custom click behavior to each activity.                                                                                       ...
+| [`17.chat-send-history`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/17.chat-send-history)                                                                                                               | Advanced tutorial: Demonstrates the ability to save user input and allow the user to step back through previous sent messages.                                           ...
+| [`18.customization-open-url`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/18.customization-open-url)                                                                                                     | Advanced tutorial: Demonstrates how to customize the open URL behavior.                                                                                                  ...
+| [`19.customization-plain-ui`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/19.customization-plain-ui)                                                                                                     | Advanced tutorial: Demonstrates how to customize Web Chat UI by building from ground up.                                                                                 ...
```
```suggestion
| [`19.customization-plain-ui`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/19.customization-plain-ui)                                                                                                     | Advanced tutorial: Demonstrates how to customize the Web Chat UI by building from ground up instead of needing to rewrite entire Web Chat components.                                                                                                                                       | [Plain UI Demo](https://microsoft.github.io/BotFramework-WebChat/19.customization-plain-ui)                               |
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+Below is the explanation of different layers of Web Chat.
+
+- [botframework-directlinejs](https://github.com/Microsoft/BotFramework-DirectLineJS) should be eventful and stateless
+- [botframework-webchat-core](https://github.com/Microsoft/BotFramework-WebChat/tree/master/packages/core) (a.k.a. Web Chat Core) is using Redux, provide stateful data store but minimal to support various UI frameworks and app scenarios
```
```suggestion
- [botframework-webchat-core](https://github.com/Microsoft/BotFramework-WebChat/tree/master/packages/core) (a.k.a. Web Chat Core) is using Redux, provides a stateful data store but minimal support to various UI frameworks and app scenarios
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+
+Below is the explanation of different layers of Web Chat.
+
+- [botframework-directlinejs](https://github.com/Microsoft/BotFramework-DirectLineJS) should be eventful and stateless
```
should talk about how it's an API to the Bot Framework protocol.

compulim replied:
Added more details on `botframework-directlinejs`.

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+
+This sample shows how to use Web Chat without its canned UI component.
+
+When we design Web Chat, we think deeply about customizations. To make customizations great, we design our component in layers. Developers can go into the full experience of Web Chat by using all of our layers as-is. Or they could build their own UI from ground up with our business logic. Or they can opt into our UIs but replacing just a handful of components.
```
```suggestion
When we designed Web Chat, we heavily considered the importance of customization. To make customization great and accessible, we designed our components in layers. Developers can get the full experience of Web Chat by:
1. using all of our layers as-is,
2. building their own UI from the ground up using Web Chat's business logic, or
3. opting into our UI but replacing just a handful of components as needed.
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+   - [`Composer`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/Composer.js) is the React context which connect to Web Chat Core. If you want to build a component using Web Chat, you will need to put the component your inside `<Composer>`. This is very similar to Redux `<Provider>` but using React Context instead
+   - [`connectToWebChat`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/connectToWebChat.js) should provide mechanisms to build components from ground up to support various creative ideas. `connectToWebChat` is very similar to Redux `connect` function, provides access to all data required to build a chat component
+   - [`MicrophoneButton.connectMicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) should provide HOC function to hoist features from Web Chat as props to the connected component. Although this is very similar to `connectToWebChat`, it provide encapsulation of information and business logics required for the specific ...
+   - [`MicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) is the actual component developer can use to reproduce the full Web Chat experience
```
Lines 9-16 might work better as a table or something.

compulim replied:
Converted to table format.

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+- [botframework-webchat-core](https://github.com/Microsoft/BotFramework-WebChat/tree/master/packages/core) (a.k.a. Web Chat Core) is using Redux, provide stateful data store but minimal to support various UI frameworks and app scenarios
+   - UI framework agnostic: we think about building apps on React Native and Angular (using Redux bindings)
+- [botframework-webchat-component](https://github.com/Microsoft/BotFramework-WebChat/tree/master/packages/component) (a.k.a. Web Chat UI) has multiple layers
+   - [`Composer`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/Composer.js) is the React context which connect to Web Chat Core. If you want to build a component using Web Chat, you will need to put the component your inside `<Composer>`. This is very similar to Redux `<Provider>` but using React Context instead
```
```suggestion
   - [`Composer`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/Composer.js) is the React context which connects to Web Chat Core. If you want to build a component using Web Chat, you will need to add the component inside `<Composer>`. This is very similar to Redux `<Provider>` but using React Context instead
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+   - UI framework agnostic: we think about building apps on React Native and Angular (using Redux bindings)
+- [botframework-webchat-component](https://github.com/Microsoft/BotFramework-WebChat/tree/master/packages/component) (a.k.a. Web Chat UI) has multiple layers
+   - [`Composer`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/Composer.js) is the React context which connect to Web Chat Core. If you want to build a component using Web Chat, you will need to put the component your inside `<Composer>`. This is very similar to Redux `<Provider>` but using React Context instead
+   - [`connectToWebChat`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/connectToWebChat.js) should provide mechanisms to build components from ground up to support various creative ideas. `connectToWebChat` is very similar to Redux `connect` function, provides access to all data required to build a chat component
```
```suggestion
   - [`connectToWebChat`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/connectToWebChat.js) provides the mechanisms to build components from the ground up to support various creative ideas, such as ______ or _______. The `connectToWebChat` method is very similar to Redux's `connect` function, and provides access to all of the data required to build a chat component
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+- [botframework-webchat-component](https://github.com/Microsoft/BotFramework-WebChat/tree/master/packages/component) (a.k.a. Web Chat UI) has multiple layers
+   - [`Composer`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/Composer.js) is the React context which connect to Web Chat Core. If you want to build a component using Web Chat, you will need to put the component your inside `<Composer>`. This is very similar to Redux `<Provider>` but using React Context instead
+   - [`connectToWebChat`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/connectToWebChat.js) should provide mechanisms to build components from ground up to support various creative ideas. `connectToWebChat` is very similar to Redux `connect` function, provides access to all data required to build a chat component
+   - [`MicrophoneButton.connectMicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) should provide HOC function to hoist features from Web Chat as props to the connected component. Although this is very similar to `connectToWebChat`, it provide encapsulation of information and business logics required for the specific ...
```
```suggestion
   - [`MicrophoneButton.connectMicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) provides HOC function to hoist features from Web Chat as props to the connected component. Although this is very similar to `connectToWebChat`, it provides encapsulation of information and the business logic required for the specific component
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+   - [`Composer`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/Composer.js) is the React context which connect to Web Chat Core. If you want to build a component using Web Chat, you will need to put the component your inside `<Composer>`. This is very similar to Redux `<Provider>` but using React Context instead
+   - [`connectToWebChat`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/connectToWebChat.js) should provide mechanisms to build components from ground up to support various creative ideas. `connectToWebChat` is very similar to Redux `connect` function, provides access to all data required to build a chat component
+   - [`MicrophoneButton.connectMicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) should provide HOC function to hoist features from Web Chat as props to the connected component. Although this is very similar to `connectToWebChat`, it provide encapsulation of information and business logics required for the specific ...
+   - [`MicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) is the actual component developer can use to reproduce the full Web Chat experience
```
```suggestion
   - [`MicrophoneButton`](https://github.com/Microsoft/BotFramework-WebChat/blob/master/packages/component/src/SendBox/MicrophoneButton.js) is the actual component the developer can use to reproduce the full Web Chat experience
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+- Type `image` in the send box
+   - It should show an image
+- Type `suggested-actions` in the send box
+   - `imBack`, `messageBack`, and `postBack` button should work as expected
```
```suggestion
   - `imBack`, `messageBack`, and `postBack` buttons should work as expected
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+
+This project has two goals:
+
+- Using Web Chat without using any of its UI component
```
```suggestion
- Use Web Chat without using any of its UI components
```

corinagum commented on samples/19.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,53 @@
+   - Support `imBack`, `messageBack`, and `postBack`
+   - Support rendering hyperlinks
+
+Although UI is main part of Web Chat, the state machine and business logic in Web Chat core is very valuable for building your highly-customized chat UI.
```
```suggestion
Although the UI is the main part of Web Chat, the state machine and business logic in Web Chat Core is very valuable for building your own highly-customized chat UI.
```

corinagum commented on samples/19.customization-plain-ui/src/App.js:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  return (
+    // We are using "Composer" component here, which all descendants will have access to Web Chat API by HOC-ing thru "connectToWebChat".
```
```suggestion
    // We are using the "Composer" component here, which all descendants will have access to the Web Chat API by HOC-ing thru "connectToWebChat".
```

corinagum commented on samples/19.customization-plain-ui/src/CardActionButton.js:
```diff
@@ -0,0 +1,19 @@
+import MessageBackButton from './MessageBackButton';
+import PostBackButton from './PostBackButton';
+
+// "CardAction" could be either, "imBack", "messageBack", or "postBack".
```
```suggestion
// "cardAction" could be either, "imBack", "messageBack", or "postBack".
```

corinagum commented on samples/19.customization-plain-ui/src/MessageBackButton.js:
```diff
@@ -0,0 +1,16 @@
+
+const MessageBackButton = ({ cardAction, sendMessageBack }) =>
+  <button
+    // Web Chat do the heavylifting for us by exposing a "sendMessageBack" function.
```
```suggestion
    // Web Chat does the heavylifting for us by exposing a "sendMessageBack" function.
```

corinagum commented on samples/19.customization-plain-ui/src/PlainWebChat.js:
```diff
@@ -0,0 +1,86 @@
+    <div>
+      <ul>
+        { activities
+            // Currently, we only display activity of type "message"
```
```suggestion
            // Currently, this sample only displays an activity of type "message"
```

corinagum commented on samples/19.customization-plain-ui/src/PlainWebChat.js:
```diff
@@ -0,0 +1,86 @@
+            .filter(({ type }) => type === 'message')
+            // We need to hide "postBack" message sent by the user
+            .filter(({ channelData: { postBack } = {}, from: { role } }) => !(role === 'user' && postBack))
+            // We normalize the activity:
```
```suggestion
            // Normalize the activity:
```

corinagum commented on samples/19.customization-plain-ui/src/PlainWebChat.js:
```diff
@@ -0,0 +1,86 @@
+            // We need to hide "postBack" message sent by the user
+            .filter(({ channelData: { postBack } = {}, from: { role } }) => !(role === 'user' && postBack))
+            // We normalize the activity:
+            // - Every activity should have an "attachments" array, consists of zero or more attachments;
```
```suggestion
            // - Every activity should have an "attachments" array, consisting of zero or more attachments:
```

corinagum commented on samples/19.customization-plain-ui/src/PlainWebChat.js:
```diff
@@ -0,0 +1,86 @@
+            // We normalize the activity:
+            // - Every activity should have an "attachments" array, consists of zero or more attachments;
+            // - If this is a "messageBack" message, we should use the "displayText",
+            //   because "text" is being submitted to bot, and "displayText" is what we use to override the bot displayed to the user.
```
```suggestion
            //   because "text" is being submitted to bot, and "displayText" is what we use to override what the bot displays
            
             
              
               
                
                 
                  
                    to the user.
```

corinagum commented on samples/19.customization-plain-ui/src/SuggestedActions.js:
```diff
@@ -0,0 +1,18 @@
+import CardActionButton from './CardActionButton';
+
+// Web Chat cleaned up the suggestedActions for us.
+// If the last activity is from bot and contains "suggestedActions", Web Chat will send it to us thru "suggestedActions".
```
```suggestion
// If the last activity is from the bot and contains "suggestedActions", Web Chat will send it to us thru "suggestedActions".
```

corinagum commented:
I think this README is a good explanation for customizing Web Chat.  If you want to highlight this as a good tutorial for customization, I recommend adding emphasis on the root README.md and SAMPLES.md table. We might also benefit from a general FAQ section as well.

compulim commented:
@corinagum I have updated the PR with changes for `README.md`. Code is mainly untouched but added a few paragraphs talking about the purpose of this sample (rather than a blank white screen).

compulim commented:
@corinagum could you look at this one again?

corinagum commented on samples/21.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,157 @@
+
+1. Using all of our layers: using all of our layers as-is,
+2. Using just business layers: building their own UI from the ground up using Web Chat's business logic, or
+3. Using business layers and partly UI components: opting into our UI but replacing just a handful of components as needed.
```
```suggestion
3. Using business layers and some UI components: opting into our UI but replacing just a handful of components as needed.
```

corinagum commented on samples/21.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,157 @@
+               <li>
+                  Designed to be event-driven (via observables) and stateless
+                  <ul>
+                     <li>Due to its stateless nature, for example, it does not understood suggested actions, because suggested actions depends on previous state</li>
```
```suggestion
                     <li>Due to its stateless nature, for example, it does not understand suggested actions, because suggested actions depends on previous state</li>
```

corinagum commented on samples/21.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,157 @@
+
+Below is the explanation of different layers of Web Chat.
+
+<table>
```
I feel like this could be super beneficial to new users in the root README.md. Should we consider adding a link to this page so people can learn more about the components?

compulim replied:
Great idea.

But since our sample is scattered (or littered) in `SAMPLES.md`, instead of copying to `SAMPLES.md` for now. I would prefer cleaning up `SAMPLES.md` or building a GitHub Wiki out of it.

corinagum commented on samples/21.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,157 @@
+            </a>
+         </td>
+         <td>
+           <a href="https://reactjs.org/docs/higher-order-components.html">Higher-order component</a> to expose all Web Chat UI API as props. The <code>connectToWebChat</code> method is very similar to Redux <a href="https://react-redux.js.org/api/connect"><code>connect</code></a> function, and provides access to all of the data required to build a chat component.
```
```suggestion
           <a href="https://reactjs.org/docs/higher-order-components.html">Higher-order component</a> that exposes all of Web Chat's UI API as props. The <code>connectToWebChat</code> method is very similar to Redux <a href="https://react-redux.js.org/api/connect"><code>connect</code></a> function, and provides access to all of the data required to build a chat component.
```

corinagum commented on samples/21.customization-plain-ui/README.md:
```diff
@@ -0,0 +1,157 @@
+
+This project has two goals:
+
+-  Use Web Chat without using any of its UI components
```
```suggestion
-  Use Web Chat without using any of its original UI components
```

compulim commented on README.md:
```diff
@@ -191,6 +191,7 @@ See more about [customizing Web Chat](https://github.com/Microsoft/BotFramework-
 | [`17.chat-send-history`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/17.chat-send-history)                                                                                                               | Advanced tutorial: Demonstrates the ability to save user input and allow the user to step back through previous sent messages.                                           ...
 | [`18.customization-open-url`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/18.customization-open-url)                                                                                                     | Advanced tutorial: Demonstrates how to customize the open URL behavior.                                                                                                  ...
 | [`19.a.single-sign-on-for-enterprise-apps`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/19.a.single-sign-on-for-enterprise-apps)                                                                         | Demonstrates how to use single sign-on for enterprise apps using OAuth                                                                                                   ...
+| [`21.customization-plain-ui`](https://github.com/Microsoft/BotFramework-WebChat/tree/master/samples/21.customization-plain-ui)                                                                                                     | Advanced tutorial: Demonstrates how to customize the Web Chat UI by building from ground up instead of needing to rewrite entire Web Chat components.                    ...
```
Added to `SAMPLES.md`.
